### PR TITLE
[Do not merge] JUnit 5 migration

### PR DESCRIPTION
**Pull request by tonidero**
### Motivation
https://revenuecats.atlassian.net/browse/CSDK-48

Junit5 has some advantages for tests like parameterized tests or better test organization. This PR is a PoC of the migration to Junit 5

### Description
During this PoC, I discovered Robolectric is not compatible with Junit5. Since we use a bunch of classes from the Android sdk, this makes it difficult for the tests to run normally. While some classes can be mocked relatively easily, others are harder to mock. The main component that would be difficult to mock is the json library. This library is included in the android sdk and is used in several places in the sdk.

In this PR, I tried importing the official json library for the tests, so we wouldn't have to mock it. However, I noticed there are differences in behavior when compared to the version inside the android sdk, which would be no good since the same code might behave different in production and test code.

I also had to mock a few android classes and fix a few tests.

### CODE REVIEW

File: build.gradle [hunk 1]
```diff
@@ -15,6 +16,7 @@ buildscript {
         classpath 'com.android.tools.build:gradle:7.0.2'
         classpath "org.jetbrains.kotlin:kotlin-gradle-plugin:$kotlinVersion"
         classpath "org.jetbrains.dokka:dokka-gradle-plugin:1.6.10"
+        classpath "de.mannodermaus.gradle.plugins:android-junit5:1.8.2.0"
```
Comment (tonidero) [inline build.gradle:19]: This plugin configures test tasks to execute junit5 correctly. We could try not to add this dependency, but considering it was only going to be used for tests and it was popular enough, I thought it was ok.

File: common/build.gradle [hunk 2]
```diff
@@ -13,12 +17,13 @@ dependencies {
     testImplementation 'androidx.test:core:1.2.0'
     testImplementation 'androidx.test:runner:1.2.0'
     testImplementation 'androidx.test:rules:1.2.0'
-    testImplementation 'androidx.test.ext:junit:1.1.1'
+    testImplementation "org.junit.jupiter:junit-jupiter-api:$junit5Version"
+    testRuntimeOnly "org.junit.jupiter:junit-jupiter-engine:$junit5Version"
+    testImplementation 'org.json:json:20210307'
```
Comment (tonidero) [inline common/build.gradle:22]: This is adding the latest official json library to the tests of the `common` module. As mentioned We have to either do this, or mock all json related classes we were using. In both cases, we would lose the ability to test with the real implementation of those classes (the one included on the android sdk)

File: common/src/main/java/com/revenuecat/purchases/common/Backend.kt [hunk 3]
```diff
@@ -292,10 +292,6 @@ class Backend(
         }
     }
 
-    private fun encode(string: String): String {
-        return Uri.encode(string)
```
Comment (tonidero) [inline common/src/main/java/com/revenuecat/purchases/common/Backend.kt:296]: Moved this functionality to an external class to make it easier to mock

File: common/src/test/java/com/revenuecat/purchases/common/AppConfigTest.kt [hunk 4]
```diff
@@ -1,18 +1,18 @@
...
 import io.mockk.mockkStatic
 import org.assertj.core.api.Assertions.assertThat
-import org.junit.Test
-import org.junit.runner.RunWith
+import org.junit.jupiter.api.Test
+import org.junit.jupiter.api.extension.ExtendWith
 import java.net.URL
 
-@RunWith(AndroidJUnit4::class)
+@ExtendWith(LogMockExtension::class)
```
Comment (tonidero) [inline common/src/test/java/com/revenuecat/purchases/common/AppConfigTest.kt:15]: This allows to extend the tests with some custom code. This extension (you can see the code below) mocks the `Log` static methods which are part of the android sdk

File: common/src/test/java/com/revenuecat/purchases/common/OfferingsTest.kt [hunk 5]
```diff
@@ -129,7 +124,7 @@ class OfferingsTest {
     @Test
     fun `No offerings`() {
         val offerings =
-            JSONObject("{'offerings': [], 'current_offering_id': null}").createOfferings(emptyMap())
+            JSONObject("{'offerings': [], 'current_offering_id': ''}").createOfferings(emptyMap())
```
Comment (tonidero) [inline common/src/test/java/com/revenuecat/purchases/common/OfferingsTest.kt:127]: This test is what made me reconsider this approach... Looks like `JSONObject.getString` in the android sdk with a null value would work. However, that same method on the official json library would throw an exception saying that the value is not a string... When considering that difference, we could potentially have tests that differ in behavior from their production code... So I considered this approach is problematic

File: library.gradle [hunk 6]
```diff
@@ -49,8 +49,6 @@ dependencies {
     testImplementation 'androidx.test:rules:1.4.0'
     testImplementation 'androidx.test.ext:junit:1.1.3'
     testImplementation 'org.robolectric:robolectric:4.6.1'
-    testImplementation 'com.squareup.okhttp3:mockwebserver:4.2.0'
-    testImplementation 'org.mockito:mockito-core:3.0.0'
```
Comment (tonidero) [inline library.gradle:53]: These are a bunch of libraries we are not using... Looks like we are not using mockito anywhere, and `mockwebserver` only in one of the modules.

File: utils/build.gradle [hunk 7]
```diff
@@ -4,7 +4,4 @@ dependencies {
     implementation fileTree(dir: "libs", include: ["*.jar"])
     implementation "org.jetbrains.kotlin:kotlin-stdlib:$kotlinVersion"
     implementation 'androidx.core:core-ktx:1.3.1'
-    testImplementation 'junit:junit:4.12'
-    androidTestImplementation 'androidx.test.ext:junit:1.1.1'
-    androidTestImplementation 'androidx.test.espresso:espresso-core:3.2.0'
```
Comment (tonidero) [inline utils/build.gradle:9]: There are no tests on the `utils` module so these are unneeded

**Comment by tonidero** [comment]
Closing this following the conversation in https://revenuecats.atlassian.net/browse/CSDK-48

File: utils/build.gradle (hunk 7 above)
Comment (vegaro) [reply utils/build.gradle:9]: Same for these, I also noticed them when I was trying to make the compilation faster. Do you mind opening a PR?

Comment (vegaro) [reply utils/build.gradle:9]: Or create a task in Jira, it's not a high priority 😄

**Pull request closed without merging**